File: Best Time to Buy and Sell Stock.py
class Solution:
	# @param prices, a list of integer
	# @return an integer
	def maxProfit(self, prices):
		if (not prices):
			return 0
		payment = -prices[0]
		profit = 0
		for price in prices:
			payment = max(payment, -price)
			profit = max(profit, price + payment)
		return profit

File: Best Time to Buy and Sell Stock.java
public class Solution {
    public int maxProfit(int[] prices) {
        if (prices.length == 0)
            return 0;
        int minPay = prices[0];
        int maxProfit = 0;
        for (int i = 1; i < prices.length; i++) {
            if (prices[i] < minPay)
                minPay = prices[i];
            if (prices[i] - minPay > maxProfit)
                maxProfit = prices[i] - minPay;
        }
        return maxProfit;
    }
}